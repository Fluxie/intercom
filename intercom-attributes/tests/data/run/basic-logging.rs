extern crate intercom;
extern crate log;
extern crate simple_logger;

use submodule::ISubInterface;

#[intercom::com_class(IInterface, ISubInterface)]
struct S;

#[intercom::com_interface]
trait IInterface
{
    fn call(&self);
}

#[intercom::com_interface]
trait IAnotherInterface
{
}

#[intercom::com_impl]
impl IInterface for S
{
    fn call(&self)
    {
        println!("Call");
    }
}

mod submodule
{
    use super::*;

    #[intercom::com_interface]
    pub trait ISubInterface
    {
    }

    #[intercom::com_impl]
    impl ISubInterface for S {}
}

fn main()
{
    simple_logger::init().unwrap();

    log::info!("Acquire S as IInterface");
    let combox = intercom::ComBox::new(S);
    let rc: intercom::ComRc<dyn IInterface> = intercom::ComRc::from(combox);

    log::info!("Call IInterface::call");
    rc.call();

    log::info!("Query ISubInterface");
    let _: intercom::ComRc<dyn submodule::ISubInterface> =
        intercom::ComItf::query_interface(&rc).unwrap();

    log::info!("Query IAnotherInterface");
    let rc: intercom::ComResult<intercom::ComRc<dyn IAnotherInterface>> =
        intercom::ComItf::query_interface(&rc);

    log::info!("Cleanup");
}
