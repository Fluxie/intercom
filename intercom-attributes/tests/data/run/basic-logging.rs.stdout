<TIMESTAMP> INFO  [testcrate] Acquire S as IInterface
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(21495B79-90CF-33C7-7032-9BC7C31AD5F9)
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(21495B79-90CF-33C7-7032-9BC7C31AD5F9) -> IInterface (Automation) [xxxxxxxx]
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(C96D83CC-F4BC-3B80-6951-3692BB4FDBC7)
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(C96D83CC-F4BC-3B80-6951-3692BB4FDBC7) -> IInterface (Raw) [xxxxxxxx]
<TIMESTAMP> INFO  [testcrate] Call IInterface::call
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] Calling IInterface::call
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] Calling IInterface::call, type system: Automation
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] Serving testcrate::S::call
Call
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] Serving testcrate::S::call, OK
<TIMESTAMP> INFO  [testcrate] Query ISubInterface
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface, type system: Automation
<TIMESTAMP> TRACE [intercom::interfaces] [xxxxxxxx] Serving testcrate::S::query_interface
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(4CC91049-3D9A-3498-46FF-932AEDA28A33)
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(4CC91049-3D9A-3498-46FF-932AEDA28A33) -> ISubInterface (Raw) [xxxxxxxx]
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::release
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::release, type system: Raw
<TIMESTAMP> TRACE [intercom::interfaces] [xxxxxxxx] Serving testcrate::S::release
<TIMESTAMP> INFO  [testcrate] Query IAnotherInterface
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface, type system: Automation
<TIMESTAMP> TRACE [intercom::interfaces] [xxxxxxxx] Serving testcrate::S::query_interface
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(46B4908C-8685-327F-7766-42C217D7984B)
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(46B4908C-8685-327F-7766-42C217D7984B) -> E_NOINTERFACE
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::query_interface, type system: Automation
<TIMESTAMP> TRACE [intercom::interfaces] [xxxxxxxx] Serving testcrate::S::query_interface
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(25A761EE-99D6-3A40-57CF-7F6A4ACD1615)
<TIMESTAMP> TRACE [generated::testcrate] [xxxxxxxx] S::query_interface(25A761EE-99D6-3A40-57CF-7F6A4ACD1615) -> E_NOINTERFACE
<TIMESTAMP> INFO  [testcrate] Cleanup
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::release
<TIMESTAMP> TRACE [generated::intercom::interfaces] [xxxxxxxx] Calling RawIUnknown::release, type system: Automation
<TIMESTAMP> TRACE [intercom::interfaces] [xxxxxxxx] Serving testcrate::S::release
